presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/IndexedPriorityQueue.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.resourceGroups;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A priority queue with constant time contains(E) and log time remove(E)
+ * Ties are broken by insertion order
+ */
+final class IndexedPriorityQueue<E>
+        implements UpdateablePriorityQueue<E>
+{
+    private final Map<E, Entry<E>> index = new HashMap<>();
+    private final Set<Entry<E>> queue = new TreeSet<>((entry1, entry2) -> {
```

Comment:
Because I wouldn't be including `value` in the comparison, and if I implemented `equals` it would definitely include `value`, so that would lead to very confusing results